R/aaa_survival_prob.R

```diff
@@ -145,3 +145,142 @@ matrix_to_nested_tibbles_survival <- function(x, time) {
   dplyr::group_nest(res, .row, .key = ".pred")$.pred
 }
 
+
+
+# summary_survfit helpers -------------------------------------------------
+
+
+survfit_summary_typestable <- function(object){
+  # make matrix of dimension n_times x n_obs
+  sanitize_element <- function(x) {
+    if (!is.matrix(x)) {
+      x <- matrix(x, ncol = length(object$n))
+    }
+    x
+  }
+  # sanitize elements we care about
+  elements <- available_survfit_summary_elements(object)
+  for (i in elements) {
+    object[[i]] <- sanitize_element(object[[i]])
+  }
+
+  object
+}
+
+available_survfit_summary_elements <- function(object) {
+  intersect(
+    names(object),
+    c("surv", "std.err", "lower", "upper", "cumhaz", "std.chaz")
+  )
+}
+
+survfit_summary_patch_infinite_time <- function(object, time) {
+
+  time_neg_inf <- is.infinite(time) & (time < 0)
+  time_inf <- is.infinite(time) & (time > 0)
+
+  patch_neg_inf <- function(x, value) {
+    rbind(
+      matrix(value, nrow = sum(time_neg_inf), ncol = ncol(x)),
+      x
+    )
+  }
+  patch_inf <- function(x, value) {
+    rbind(
+      x,
+      matrix(value, nrow = sum(time_inf), ncol = ncol(x))
+    )
+  }
+
+  # glmnet does not provide standard errors etc
+  has_std_error <- "std.err" %in% names(object)
+
+  if (any(time_neg_inf)) {
+    object$surv <- patch_neg_inf(object$surv, value = 1)
+    object$cumhaz <- patch_neg_inf(object$cumhaz, value = 0)
+    if (has_std_error) {
+      object$std.err <- patch_neg_inf(object$std.err, value = NA_real_)
+      object$lower <- patch_neg_inf(object$lower, value = NA_real_)
+      object$upper <- patch_neg_inf(object$upper, value = NA_real_)
+      object$std.chaz <- patch_neg_inf(object$std.chaz, value = NA_real_)
+    }
+  }
+  if (any(time_inf)) {
+    object$surv <- patch_inf(object$surv, value = 0)
+    object$cumhaz <- patch_inf(object$cumhaz, value = 1)
+    if (has_std_error) {
+      object$std.err <- patch_inf(object$std.err, value = NA_real_)
+      object$lower <- patch_inf(object$lower, value = NA_real_)
+      object$upper <- patch_inf(object$upper, value = NA_real_)
+      object$std.chaz <- patch_inf(object$std.chaz, value = NA_real_)
+    }
+  }
+
+  object
+}
+
+survfit_summary_restore_time_order <- function(object, time) {
+  # preserve original order of `time` because `summary()` returns a result for
+  # an ordered vector of finite time
+  # Note that this requires a survfit summary object which has already been
+  # patched for infinite time points
+  original_order_time <- match(time, sort(time))
+
+  elements <- available_survfit_summary_elements(object)
+
+  # restore original order of prediction time points
+  for (i in elements) {
+    object[[i]] <- object[[i]][original_order_time, , drop = FALSE]
+  }
+
+  object
+}
+
+survfit_summary_patch_missings <- function(object, index_missing, time, n_obs) {
+  if (is.null(index_missing)) {
+    return(object)
+  }
+
+  patch_element <- function(x) {
```

Comment:
This probably isn't going to be an issue but relying on scoping rules to find the right data is fairly problematic. I usually suggest adding arguments for anything that is used in the function.